pkg/chunk/delete_store.go

```diff
@@ -0,0 +1,366 @@
+package chunk
+
+import (
+	"context"
+	"encoding/binary"
+	"encoding/hex"
+	"flag"
+	"fmt"
+	"hash/fnv"
+	"strings"
+	"time"
+
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/labels"
+)
+
+const (
+	// DeleteRequestStatusReceived means no actions done on request yet, just doing query time filtering
+	DeleteRequestStatusReceived = "0"
+	// DeleteRequestStatusBuildingPlan means request is picked up for processing and building plans for it, still doing query time filtering
+	DeleteRequestStatusBuildingPlan = "1"
+	// DeleteRequestStatusDeleting means plans built already, running delete operations and still doing query time filtering
+	DeleteRequestStatusDeleting = "2"
+	// DeleteRequestStatusProcessed means all requested data deleted, not considering this for query time filtering
+	DeleteRequestStatusProcessed = "3"
+
+	// CacheKindStore is for cache gen number for store cache
+	CacheKindStore = "0"
+	// CacheKindResults is for cache gen number for results cache
+	CacheKindResults = "1"
+)
+
+var pendingDeleteRequestStatuses = []string{DeleteRequestStatusReceived, DeleteRequestStatusBuildingPlan, DeleteRequestStatusDeleting}
+
+// DeleteRequest holds all the details about a delete request
+type DeleteRequest struct {
+	RequestID string              `json:"request_id"`
+	UserID    string              `json:"-"`
+	StartTime model.Time          `json:"start_time"`
+	EndTime   model.Time          `json:"end_time"`
+	Selectors []string            `json:"selectors"`
+	Status    string              `json:"status"`
+	Matchers  [][]*labels.Matcher `json:"-"`
+	CreatedAt model.Time          `json:"created_at"`
+}
+
+// DeleteStore provides all the methods required to manage lifecycle of delete request and things related to it
+type DeleteStore interface {
+	Add(ctx context.Context, userID string, startTime, endTime model.Time, selectors []string) error
```

Comment:
This function should be called `AddDeletePlan` or `CreateDeletePlan`

Reply:
`Add` function adds a new delete request so I guess you mean `AddDeleteRequest`